pokedex/model/csrc/pokedex_interface.c

```diff
@@ -207,34 +234,126 @@ static void* pokedex_create(
 
 static void pokedex_destroy(void* _model) {
     if (_model) {
-        instance_mutex = 0;
+        atomic_flag_clear_explicit(&instance_mutex, memory_order_release);
     }
 }
 
 static void pokedex_reset(void* _model, uint32_t initial_pc) {
     (void)(_model);
 
+    trace_buffer.valid = 0;
     ASL_ResetState_0();
     PC_write_0(initial_pc);
 }
 
-static void pokedex_step(void* _model) {
+static uint8_t pokedex_step_trace(
+    void* _model,
+    const struct pokedex_mem_callback_vtable* mem_callback_vtable,
+    void* mem_callback_data
+) {
     (void)(_model);
 
-    ASL_Step_0();
+    mem_cb_vtable = mem_callback_vtable;
+    mem_cb_data = mem_callback_data;
+
+    memset(&trace_buffer, 0, sizeof(trace_buffer));
+    trace_buffer.valid  = 1;
+    trace_buffer.pc = PC_read_0();
+
+    FFI_StepResult result = ASL_Step_0();
+
+    trace_buffer.step_status = result.code;
+    trace_buffer.inst = result.inst;
+
+    mem_cb_vtable = NULL;
+    mem_cb_data = NULL;
+
+    return result.code;
 }
 
-static struct pokedex_vtable vtable = {
-    .abi_version = POKEDEX_ABI_VERSION,
+const struct pokedex_trace_buffer* pokedex_get_trace_buffer(void* _model) {
+    (void)(_model);
+
+    return &trace_buffer;
+}
+
+static uint64_t sext(uint32_t x) {
+    return (uint64_t)(int64_t)(int32_t)x;
+}
+
+static uint64_t nanbox(uint32_t x) {
+    return 0xFFFFFFFF00000000 | x;
+}
+
+static void pokedex_read_pc(void* _model, uint64_t* ret){
+    (void)(_model);
+
+    *ret = sext(ASL_read_PC_0());
+}
+
+static void pokedex_read_xreg(void* _model, uint8_t xs, uint64_t* ret){
+    (void)(_model);
+
+    *ret = sext(ASL_read_XREG_0(xs));
+}
+
+static void pokedex_read_freg(void* _model, uint8_t fs, uint64_t* ret){
+    (void)(_model);
+
+    *ret = nanbox(ASL_read_FREG_0(fs));
+}
+
+static void pokedex_read_vreg(void* _model, uint8_t vs, uint8_t* buf, size_t buflen) {
+    (void)(_model);
+
+    // FIXME: get VLEN from ASL side instead of hard-coding
+    const int VLEN = 256;
+
+    assert(buflen == VLEN / 8);
+
+    unsigned _BitInt(256) vreg = ASL_read_VREG_0(vs);
+    memcpy(buf, &vreg, VLEN / 8);
+}
+
+static void pokedex_read_csr(void* _model, uint16_t csr, uint64_t* ret){
+    (void)(_model);
+
+    // zero extension
+    *ret = ASL_read_CSR_0(csr);
+}
+
+static const struct pokedex_model_description model_desc = {
     .model_isa = "rv32imacf_zve32_zvl256",
+    .xlen = 32,
+    .flen = 32,
+    .vlen = 256,
+};
+
+const struct pokedex_model_description* pokedex_get_description(void* _model) {
+    (void)(_model);
+
+    return &model_desc;
+}
+
+static const struct pokedex_export export = {
+    .abi_version = POKEDEX_ABI_VERSION,
     .create = pokedex_create,
     .destroy = pokedex_destroy,
 
+    .get_description = pokedex_get_description,
+
     .reset = pokedex_reset,
-    .step = pokedex_step,
+    .step = pokedex_step_trace,
```

Comment:
Do we need to keep `step` or `step` is for future design?

Reply:
Yes, actually `step` is the primary method. When we are not doing difftest, `step` is enough, e.g. use pokedex only to run testcases, or act as an gdb stub when doing interactive debugging.

`step_trace` is an optimization to speed up difftest. The caller could simply call `step` then read everything through accessors.

In our current implementation, since ASL is too slow, tracing adds little overhead, so no bother to write a dedicated "step without trace". But I'd like make the distinction clear in concept.